pineappl/src/boc.rs

```diff
@@ -209,6 +212,460 @@ impl Scales {
+                let (lhs, rhs) = {
+                    if let Some((lhs, rhs)) = string.split_once(':') {
+                        (lhs.parse::<usize>(), rhs.parse::<usize>())
+                    } else {
+                        // there's no colon
+                        continue;
+                    }
+                };
+
+                if let Ok(num) = rhs {
+                    vec.truncate(vec.len() - num);
+                }
+
+                if let Ok(num) = lhs {
+                    vec.drain(0..num);
+                }
+
+                if vec.len() <= 1 {
+                    return Err(Bla::Error("no limits due to ':' syntax".to_owned()));
+                }
+            }
+        }
+
+        let dimensions = remaps.len();
+        let mut normalizations = Vec::new();
+        let mut limits = Vec::new();
+        let mut buffer = Vec::with_capacity(dimensions);
+        let mut pipe_indices = vec![0; dimensions];
+        let mut last_indices = vec![0; dimensions];
+
+        'looop: for indices in remaps
+            .iter()
+            .map(|vec| 0..vec.iter().map(|vec| vec.len() - 1).max().unwrap())
+            .multi_cartesian_product()
+        {
+            // calculate `pipe_indices`, which stores the indices for the second dimension of `remaps`
+            for d in 0..dimensions - 1 {
+                if indices[d] > last_indices[d] {
+                    for dp in d + 1..dimensions {
+                        if remaps[dp].len() != 1 {
+                            pipe_indices[dp] += 1;
+                        }
+                    }
+                }
+            }
+
+            last_indices.clone_from(&indices);
+
+            let mut normalization = 1.0;
+
+            for (remap, &pipe_index, &i) in izip!(&remaps, &pipe_indices, &indices) {
+                if let Some(r) = remap.get(pipe_index) {
+                    if r.len() <= (i + 1) {
+                        buffer.clear();
+
+                        // this index doesn't exist
+                        continue 'looop;
+                    }
+
+                    let left = r[i];
+                    let right = r[i + 1];
+
+                    buffer.push((left, right));
+                    normalization *= right - left;
+                } else {
+                    return Err(Bla::Error(
+                        "missing '|' specification: number of variants too small".to_owned(),
+                    ));
+                }
+            }
+
+            // TODO: rewrite the code to avoid `buffer.clear()`
+            limits.push(buffer.clone());
+            buffer.clear();
+            normalizations.push(normalization);
+        }
+
+        Ok(Self::from_limits_and_normalizations(limits, normalizations)
+            // TODO: implement proper error handling
+            .unwrap())
+    }
+}
+
+/// Error type returned by [`BinRemapper::from_str`]
+#[derive(Debug, Error)]
+pub enum Bla {
```

Comment:
```suggestion
pub enum BinRemapperStrError {
```
nur um das klar zu stellen: mein Standardname is 50% `Bla` und 50% `Blub` :see_no_evil:

Reply:
Yes, that still has to be changed.